File: brainbox/remove_outermost_parentheses.ts
export function removeOuterParentheses(s: string): string {
  let result = "";
  let balance = 0;

  for (const ch of s) {
    if (ch === "(") {
      if (balance > 0) result += ch; // keep only if not outermost
      balance++;
    } else {
      balance--;
      if (balance > 0) result += ch; // keep only if not outermost
    }
  }

  return result;
}

import { removeOuterParentheses } from "./removeOuterParentheses";

function test_removeOuterParentheses() {
  console.log(removeOuterParentheses("(()())(())"), "== ()()()");
  console.log(removeOuterParentheses("(()())(())(()(()))"), "== ()()()()(())");
  console.log(removeOuterParentheses("()()"), "== ");
  console.log(removeOuterParentheses("((()))"), "== (())");
  console.log(removeOuterParentheses("(()(()))"), "== ()(())");
}

test_removeOuterParentheses();

File: brainbox/remove_outermost_parentheses.py
def remove_outermost_parentheses(s: str) -> str:
    result = []
    balance = 0

    for ch in s:
        if ch == '(':
            if balance > 0:
                result.append(ch)
            balance += 1
        else:  # ch == ')'
            balance -= 1
            if balance > 0:
                result.append(ch)

    return "".join(result)


def test_remove_outermost_parentheses():
    print(remove_outermost_parentheses("(()())(())"), "== ()()()")
    print(remove_outermost_parentheses("(()())(())(()(()))"), "== ()()()()(())")
    print(remove_outermost_parentheses("()()"), "== ")
    print(remove_outermost_parentheses("((()))"), "== (())")
    print(remove_outermost_parentheses("(()(()))"), "== ()(())")

test_remove_outermost_parentheses()
